go/private/rules/cgo.bzl

```diff
@@ -35,6 +36,28 @@ load(
     "cc_transitive_headers",
     "has_cc",
 )
+load(
+    "@io_bazel_rules_go//go/platform:list.bzl",
+    "GOARCH",
+    "GOOS",
+)
+
+_WHOLE_ARCHIVE_FLAGS_BY_GOOS = {
+    goos: ("-Wl,--whole-archive", "-Wl,--no-whole-archive")
+    for goos in GOOS.keys()
+}
+_WHOLE_ARCHIVE_FLAGS_BY_GOOS.update({
+    "darwin": ("-Wl,-all_load", ""),
```

Comment:
Is this actually an OS difference or a GCC / clang difference? If it's the latter, we should check the compiler name (unless there's a better way to test for that). I think that might be in `go.cgo_tools.c_compiler_path`. After all, you can have gcc on macOS and clang on Linux.